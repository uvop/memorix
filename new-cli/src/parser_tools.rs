use nom::combinator::cut;
use nom::error::context;
use nom::{
    branch::alt,
    bytes::complete::tag,
    character::complete::{alpha1, alphanumeric1, char, multispace0, multispace1},
    combinator::{map, recognize},
    error::ParseError,
    multi::{many0, separated_list0},
    sequence::{pair, preceded, separated_pair, terminated, tuple},
    IResult,
};

pub fn indent(level: usize) -> String {
    "  ".repeat(level)
}

#[macro_export]
macro_rules! create_enum_with_const_slice {
    (
        $(#[$meta:meta])*
        $vis:vis enum $name:ident {
            $($(#[$variant_meta:meta])* $variant:ident),+ $(,)?
        }
        $const_vis:ident $const_name:ident
    ) => {
        $(#[$meta])*
        $vis enum $name {
            $($(#[$variant_meta])* $variant),+
        }

        $const_vis const $const_name: &'static [$name] = &[
            $($name::$variant),+
        ];
    }
}

pub trait FromSdl {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized;
}
pub trait ToSdl {
    fn to_sdl(&self, level: usize) -> String;
}

impl FromSdl for String {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        map(
            recognize(pair(
                alt((alpha1, tag("_"))),
                many0(alt((alphanumeric1, tag("_")))),
            )),
            |x: &str| x.to_string(),
        )(input)
    }
}

impl ToSdl for String {
    fn to_sdl(&self, _: usize) -> String {
        self.to_string()
    }
}

impl<T: FromSdl> FromSdl for Vec<T> {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        context(
            "array",
            preceded(
                tuple((char('['), multispace0)),
                cut(terminated(
                    separated_list0(multispace1, T::from_sdl),
                    preceded(multispace0, char(']')),
                )),
            ),
        )(input)
    }
}

impl<T: ToSdl> ToSdl for Vec<T> {
    fn to_sdl(&self, level: usize) -> String {
        let level_indent = indent(level + 1);
        let mut result = String::from("[\n");

        for item in self {
            result.push_str(&format!("{}{}\n", level_indent, item.to_sdl(level + 1)));
        }
        result.push_str(&format!("{}]", indent(level)));

        result
    }
}

impl<T: FromSdl> FromSdl for Vec<(String, T)> {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        context(
            "hash",
            preceded(
                char('{'),
                cut(terminated(
                    map(
                        separated_list0(
                            multispace1,
                            separated_pair(
                                preceded(multispace0, String::from_sdl),
                                cut(tuple((multispace0, char(':')))),
                                preceded(multispace0, T::from_sdl),
                            ),
                        ),
                        |tuple_vec| tuple_vec.into_iter().map(|(k, v)| (k, v)).collect(),
                    ),
                    preceded(multispace0, char('}')),
                )),
            ),
        )(input)
    }
}

impl<T: ToSdl> ToSdl for Vec<(String, T)> {
    fn to_sdl(&self, level: usize) -> String {
        let level_indent = indent(level + 1);
        let mut result = String::from("{\n");

        for (key, value) in self {
            result.push_str(&format!(
                "{}{}: {}\n",
                level_indent,
                key,
                value.to_sdl(level + 1)
            ));
        }
        result.push_str(&format!("{}}}", indent(level)));

        result
    }
}

#[macro_export]
macro_rules! impl_from_and_to_sdl_for_enum {
    ($enum_name:ident, $($variant:ident => $str_val:expr),+ $(,)?) => {
        impl FromSdl for $enum_name {
            fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(input: &'a str) -> IResult<&'a str, Self, E>
            where
                Self: Sized,
            {
                alt((
                    $(value($enum_name::$variant, tag($str_val))),+
                ))(input)
            }
        }
        impl ToSdl for $enum_name {
            fn to_sdl(&self, _: usize) -> String {
                match self {
                    $($enum_name::$variant => $str_val.to_string()),+
                }
            }
        }
    };
}

#[macro_export]
macro_rules! impl_from_and_to_sdl_for_struct {
    (
        $(#[$attr:meta])*
        $vis:vis struct $struct_name:ident {
            $(
                $vis_field:vis $field:ident: $field_type:ty
            ),+ $(,)?
        }
    ) => {
        $(#[$attr])*
        $vis struct $struct_name {
            $(
                $vis_field $field: $field_type
            ),+
        }

        impl FromSdl for $struct_name {
            fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(input: &'a str) -> IResult<&'a str, Self, E>
            where
                Self: Sized,
            {
                preceded(
                    char('{'),
                    cut(terminated(
                        permutation((
                            $(
                                {
                                    let parser = preceded(
                                        tuple((multispace0, tag(stringify!($field)), multispace0, char(':'), multispace0)),
                                        cut(impl_from_and_to_sdl_for_struct!(@from_sdl $field_type))
                                    );
                                    impl_from_and_to_sdl_for_struct!(@final_parser $field, $field_type, parser)
                                },
                            )+
                        )),
                        preceded(multispace0, char('}'))
                    ))
                )(input)
                .map(|(remaining, fields)| {
                    let ($($field),+) = fields;
                    (remaining, $struct_name { $($field),+ })
                })
            }
        }

        impl ToSdl for $struct_name {
            fn to_sdl(&self, level: usize) -> String {
                let level_indent = indent(level + 1);
                let mut result = String::from("{\n");

                $(
                    impl_from_and_to_sdl_for_struct!(@to_sdl_field self.$field, stringify!($field), $field_type, level_indent, result, level);
                )+

                result.push_str(&format!("{}}}", indent(level)));
                result
            }
        }
    };

    (@from_sdl Option<$inner:ty>) => {
        $inner::from_sdl
    };

    (@from_sdl $other:ty) => {
        <$other>::from_sdl
    };


    (@final_parser $field:ident, Option<$inner:ty>, $parser:expr) => {
        opt($parser)
    };

    (@final_parser $field:ident, $other:ty, $parser:expr) => {
        context(concat!("Missing required key \"", stringify!($field), "\""), $parser)
    };

    (@to_sdl_field $field:expr, $field_name:expr, Option<$inner:ty>, $indent:expr, $result:expr, $level:expr) => {
        if let Some(value) = &$field {
            $result.push_str(&format!(
                concat!("{}", stringify!($field_name),": {}\n"),
                $indent,
                value.to_sdl($level + 1)
            ));
        }
    };

    (@to_sdl_field $field:expr, $field_name:expr, $field_type:ty, $indent:expr, $result:expr, $level:expr) => {
        $result.push_str(&format!(
            concat!("{}", stringify!($field_name),": {}\n"),
            $indent,
            $field.to_sdl($level + 1)
        ));
    };
}
