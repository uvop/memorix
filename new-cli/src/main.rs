mod export_schemas;
mod flat_schema;
mod imports;
mod parser;
mod parser_tools;

use std::error::Error;
use std::fmt;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::PathBuf;

use nom::error::convert_error;
use nom::Finish;
use parser::{schema_to_sdl, Schema};
use parser_tools::FromSdl;

pub struct PrettyError<T>(pub T);

impl<T> From<T> for PrettyError<T> {
    fn from(v: T) -> Self {
        Self(v)
    }
}

impl<T: fmt::Display> fmt::Debug for PrettyError<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(&self.0, f)
    }
}

impl From<Box<dyn Error>> for PrettyError<String> {
    fn from(error: Box<dyn Error>) -> Self {
        PrettyError(error.to_string())
    }
}

impl<T: fmt::Debug + fmt::Display> Error for PrettyError<T> {}

impl<T: fmt::Display> fmt::Display for PrettyError<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub trait FileSystem {
    fn read_to_string(&self, path: &str) -> io::Result<String>;
    fn write_string(&self, path: &str, content: &str) -> io::Result<()>;
    fn resolve(
        &self,
        relative_file: &str,
        target_file: &str,
    ) -> Result<String, Box<dyn std::error::Error>>;
}

pub struct RealFileSystem;

impl FileSystem for RealFileSystem {
    fn read_to_string(&self, path: &str) -> io::Result<String> {
        let mut file = File::open(path)?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        Ok(contents)
    }
    fn write_string(&self, path: &str, content: &str) -> io::Result<()> {
        let mut file = std::fs::OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(path)?;
        file.write_all(content.as_bytes())?;
        Ok(())
    }
    fn resolve(
        &self,
        relative_file: &str,
        target_file: &str,
    ) -> Result<String, Box<dyn std::error::Error>> {
        let path = PathBuf::from(relative_file);
        let new_path = path
            .parent()
            .ok_or("Couldn't get parent of schema".to_string())?
            .join(target_file);
        let abs = new_path.canonicalize()?;
        Ok(abs
            .to_str()
            .ok_or("Couldn't convert path to string")?
            .to_string())
    }
}

pub fn format<F: FileSystem>(
    fs: &F,
    path: &str,
) -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
    let input = fs.read_to_string(path)?;
    let input = input.as_str();
    let (_, schema) = Schema::from_sdl(input).finish().map_err(|e| {
        let stack = format!("parser feedback:\n{}", convert_error(input, e));
        stack
    })?;
    let formatted_input = schema_to_sdl(&schema);
    fs.write_string(path, &formatted_input)?;
    Ok(())
}

pub fn codegen<F: FileSystem>(fs: &F, path: &str) -> Result<String, PrettyError<String>> {
    let parsed_schema =
        crate::imports::ImportedSchema::new(fs, path).map_err(|x| format!("{}", x))?;
    let export_schemas = crate::export_schemas::ExportSchema::new_vec(parsed_schema);
    let flat_export_schemas = export_schemas
        .into_iter()
        .map(crate::flat_schema::FlatExportSchema::new)
        .collect::<Vec<_>>();
    Ok(serde_json::to_string_pretty(&flat_export_schemas).unwrap())
}

enum Command {
    Format,
    Codegen,
}

fn main() -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 3 {
        return Err(format!("Usage: [codegen/format] <path_to_sdl_file>").into());
    }
    let command = match args[1].as_str() {
        "format" => Ok(Command::Format),
        "codegen" => Ok(Command::Codegen),
        x => Err(format!("Unknown command \"{}\"", x)),
    }?;

    let file_path = &args[2];
    let path = PathBuf::from(file_path)
        .canonicalize()
        .map_err(|_| "Couldn't get abs path for schema".to_string())?;
    let abs_file_path = path
        .to_str()
        .ok_or("Couldn't get abs path for schema".to_string())?;

    let fs = RealFileSystem {};

    match command {
        Command::Format => format(&fs, abs_file_path)?,
        Command::Codegen => {
            let code = codegen(&fs, abs_file_path)?;
            println!("{}", code);
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::{cell::RefCell, collections::HashMap, io};

    use crate::{codegen, format, FileSystem};

    pub struct MockFileSystem {
        pub files: RefCell<HashMap<String, String>>,
    }

    impl FileSystem for MockFileSystem {
        fn read_to_string(&self, path: &str) -> io::Result<String> {
            self.files
                .borrow()
                .get(path)
                .cloned()
                .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "File not found"))
        }

        fn write_string(&self, path: &str, content: &str) -> io::Result<()> {
            self.files
                .borrow_mut()
                .insert(path.to_string(), content.to_string());
            Ok(())
        }

        fn resolve(
            &self,
            _: &str,
            target_file: &str,
        ) -> Result<String, Box<dyn std::error::Error>> {
            Ok(target_file.to_string())
        }
    }

    #[test]
    fn test_parse_example_sdl() -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([
                (
                    "another-schema.memorix".to_string(),
                    r#"
Config {
  export: {
    engine: Redis( env(REDIS_URL) )
    files: [
      {
        language: typescript
        path: "memorix.generated.ts"
      }
      {
        language: python
        path: "memorix_generated.py"
      }
      {
        language: rust
        path: "memorix_generated.rs"
      }
    ]
  }
}
Type { abc: u32 }
Cache {
  number_of_messages: {
    payload: u32
    public: [get]
  }
  messages: {
    key: string
    payload: string
  }
}
"#
                    .to_string(),
                ),
                (
                    "schema.memorix".to_string(),
                    r#"
Config {
  import: [
    "another-schema.memorix"
  ]
  export: {
    engine: Redis( env(REDIS_URL) )
    files: [
      {
        language: typescript
        path: "memorix.generated.ts"
      }
      {
        language: python
        path: "memorix_generated.py"
      }
      {
        language: rust
        path: "memorix_generated.rs"
      }
    ]
  }
}

Namespace MessageService {
    Enum {
      Operation {
        START
        STOP
      }
    }
    Cache {
        message: {
            key: string
            payload: {
            id: string
            sender_id: UserId
            recipient_id: UserId
            content: string
            timestamp: u64
            }
            public: [get]
        }
    }

    PubSub {
        new_message: {
            key: UserId
            payload: {
            message_id: string
            recipient_id: UserId
            }
            public: [subscribe]
        }
    }

    Task {
        message_processing_tasks: {
            payload: {
                message_id: string
                processing_type: string
                priority: u32
            }
            queue_type: "Fifo"
        }
    }
}

NamespaceDefaults {
  cache_ttl: "3600"
}

Cache {
  user_profile: {
    key: u32
    payload: {
      id: u32
      name: string
      email: string
    }
    public: [get]
  }
  user_session: {
    key: string
    payload: {
      user_id: u32
      session_token: string
      expiry: u64
    }
    ttl: env(USER_SESSION_TTL)
  }
}
  
Namespace UserService {
  NamespaceDefaults {
    cache_ttl: "7200"
    task_queue_type: env(USER_SERVICE_TASK_QUEUE_TYPE)
  }

  Type {
    UserId: u64
  }

  Cache {
    user_profile: {
      key: UserId
      payload: {
        id: UserId
        name: string
        email: string
      }
      public: [get]
    }
    user_session: {
      key: string
      payload: {
        user_id: UserId
        session_token: string
        expiry: u64
      }
      ttl: env(USER_SESSION_TTL)
    }
  }

  PubSub {
    user_activity: {
      key: UserId
      payload: {
        user_id: UserId
        action: string
        timestamp: u64
      }
      public: [subscribe]
    }
  }

  Task {
    user_registration_tasks: {
      payload: {
        user_id: UserId
        email: string
        registration_date: u64
      }
      public: [enqueue get_len]
    }
  }

}

"#
                    .to_string(),
                ),
            ])),
        };
        insta::assert_snapshot!(codegen(&mock_fs, "schema.memorix")?);

        Ok(())
    }

    #[test]
    fn test_fail_nicely() -> Result<(), Box<dyn std::error::Error>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([(
                "schema.memorix".to_string(),
                r#"
Config {
  export: {
    engine: Redis(env(REDIS_URL))
    files: [      {
        language: typescript
        path: "memorix.generated.ts"
      }]
  }
}

Type {
  a: [u64 u32]
}

"#
                .to_string(),
            )])),
        };

        let result = codegen(&mock_fs, "schema.memorix");
        let err = result.unwrap_err();
        insta::assert_snapshot!(err.to_string());

        Ok(())
    }
    #[test]
    fn test_format() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([(
                "schema.memorix".to_string(),
                r#"
    Config {
      export: {
        engine: Redis(env(REDIS_URL))
        files: [      {
            language: typescript
            path: "memorix.generated.ts"
          }]
      }
    }
    Namespace Science {
      Cache {
        how_many_atoms: {
          payload: u32
          key: u64
        }
      }
      PubSub {
        how_many_atoms: {
          payload: u32
        }
      }
    }
    Type {
      a: [u64      ]
    }
    Namespace Rocket {
      Cache {
        launched: {
          payload: boolean
        }
      }
      Enum {
        Operation {
          START
          STOP
        }
        Color {
          RED
          GREEN
          BLUE
        }
      }
      PubSub {
        launched: {
          payload: boolean
        }
      }
    }

    "#
                .to_string(),
            )])),
        };
        format(&mock_fs, "schema.memorix")?;
        let formatted_schema = mock_fs.read_to_string("schema.memorix")?;
        insta::assert_snapshot!(formatted_schema);

        Ok(())
    }
}
