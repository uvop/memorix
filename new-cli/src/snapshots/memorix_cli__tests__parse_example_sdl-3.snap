---
source: src/main.rs
assertion_line: 420
expression: rust_code
---
#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct InlineCachePayloadMessage {
    pub id: String,
    pub sender_id: UserId,
    pub recipient_id: UserId,
    pub content: String,
    pub timestamp: u64,
}


#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct InlinePubSubPayloadNewMessage {
    pub message_id: String,
    pub recipient_id: UserId,
}


#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct InlineTaskPayloadMessageProcessingTasks {
    pub message_id: String,
    pub processing_type: String,
    pub priority: u32,
}


#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct InlineCachePayloadUserProfile {
    pub id: u32,
    pub name: String,
    pub email: String,
}


#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct InlineCachePayloadUserSession {
    pub user_id: u32,
    pub session_token: String,
    pub expiry: u64,
}


pub type abc = u32;

#[allow(non_camel_case_types, clippy::upper_case_acronyms)]
#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug,
)]
pub enum Operation {
    START,
    STOP,
}

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCacheMessageService {
    pub message: memorix_client_redis::MemorixCacheItem<String, InlineCachePayloadMessage, memorix_client_redis::Expose, memorix_client_redis::Expose, memorix_client_redis::Expose>,
}

impl MemorixCacheMessageService {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            message: memorix_client_redis::MemorixCacheItem::new(
                memorix_base.clone(),
                "message".to_string(),
                None,
            ),
        }
    }
}

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixPubSubMessageService {
    pub new_message: memorix_client_redis::MemorixPubSubItem<UserId, InlinePubSubPayloadNewMessage, memorix_client_redis::Expose, memorix_client_redis::Expose>,
}

impl MemorixPubSubMessageService {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            new_message: memorix_client_redis::MemorixPubSubItem::new(
                memorix_base.clone(),
                "new_message".to_string(),
            ),
        }
    }
}

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixTaskMessageService {
    pub message_processing_tasks: memorix_client_redis::MemorixTaskItemNoKey<InlineTaskPayloadMessageProcessingTasks, memorix_client_redis::Expose, memorix_client_redis::Expose, memorix_client_redis::Expose, memorix_client_redis::Expose>,
}

impl MemorixTaskMessageService {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            message_processing_tasks: memorix_client_redis::MemorixTaskItemNoKey::new(
                memorix_base.clone(),
                "message_processing_tasks".to_string(),
                memorix_client_redis::MemorixTaskOptions {
                    queue_type: "Fifo",
                },
            ),
        }
    }
}

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixMessageService {
    pub cache: MemorixCacheMessageService,
    pub pubsub: MemorixPubSubMessageService,
    pub task: MemorixTaskMessageService,
}

const MEMORIX_MESSAGE_SERVICE_NAMESPACE_NAME_TREE: &[&str] = &["MessageService"];

impl MemorixMessageService {
    pub fn new(
        other: memorix_client_redis::MemorixBase,
    ) -> Result<MemorixMessageService, Box<dyn std::error::Error + Sync + Send>> {
        let memorix_base = memorix_client_redis::MemorixBase::from(
            other,
            MEMORIX_MESSAGE_SERVICE_NAMESPACE_NAME_TREE
        );
        Ok(Self {
            cache: MemorixCacheMessageService::new(memorix_base.clone()),
            pubsub: MemorixPubSubMessageService::new(memorix_base.clone()),
            task: MemorixTaskMessageService::new(memorix_base.clone()),
        })
    }
}

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCache {
    pub user_profile: memorix_client_redis::MemorixCacheItem<u32, InlineCachePayloadUserProfile, memorix_client_redis::Expose, memorix_client_redis::Expose, memorix_client_redis::Expose>,
    pub user_session: memorix_client_redis::MemorixCacheItem<String, InlineCachePayloadUserSession, memorix_client_redis::Expose, memorix_client_redis::Expose, memorix_client_redis::Expose>,
    pub number_of_messages: memorix_client_redis::MemorixCacheItemNoKey<u32, memorix_client_redis::Expose, memorix_client_redis::Hide, memorix_client_redis::Hide>,
}

impl MemorixCache {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            user_profile: memorix_client_redis::MemorixCacheItem::new(
                memorix_base.clone(),
                "user_profile".to_string(),
                None,
            ),
            user_session: memorix_client_redis::MemorixCacheItem::new(
                memorix_base.clone(),
                "user_session".to_string(),
                memorix_client_redis::MemorixCacheOptions {
                    ttl: std::env::var("USER_SESSION_TTL").expect("missing environment variable USER_SESSION_TTL"),
                },
            ),
            number_of_messages: memorix_client_redis::MemorixCacheItemNoKey::new(
                memorix_base.clone(),
                "number_of_messages".to_string(),
                None,
            ),
        }
    }
}

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {
    pub MessageService: MemorixMessageService,

    pub cache: MemorixCache,
}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new() -> Result<Memorix, Box<dyn std::error::Error + Sync + Send>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            std::env::var("REDIS_URL").expect("missing environment variable REDIS_URL"),
            MEMORIX_NAMESPACE_NAME_TREE
        )
        .await?;
        Ok(Self {
            MessageService: MemorixMessageService::new(memorix_base.clone())?,

            cache: MemorixCache::new(memorix_base.clone()),
        })
    }
}
